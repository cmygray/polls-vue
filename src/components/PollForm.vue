<template>
  <v-form class="poll-form" @submit.prevent="handleSubmit">
    <v-text-field v-model="poll_title" label="Poll Title" />

    <datetime
      v-model="starts_at"
      input-id="starts_at"
      input-class="poll-form__date-time-input"
      type="datetime"
      format="yyyy-MM-dd HH:mm"
      value-zone="Asia/Seoul"
      :minute-step="10"
      :min-datetime="now"
    >
      <label slot="before" for="starts_at">Start at</label>
    </datetime>

    <datetime
      v-model="ends_at"
      class="mb-3"
      input-id="ends_at"
      input-class="poll-form__date-time-input"
      type="datetime"
      format="yyyy-MM-dd HH:mm"
      value-zone="Asia/Seoul"
      :minute-step="10"
      :min-datetime="now"
    >
      <label slot="before" for="ends_at">End at</label>
    </datetime>

    <v-text-field
      v-for="choice in choices"
      :key="choice.id"
      v-model="choice.choice_text"
      label="Choice"
    />

    <v-btn block color="primary" type="submit" :disabled="!isUpdated"
      >Save</v-btn
    >
  </v-form>
</template>

<script lang="ts">
import { Component, Emit, Prop, Vue } from 'vue-property-decorator';
import { Choice, Poll } from '@/models';

@Component
export default class PollForm extends Vue {
  @Prop()
  poll: Poll;

  now: string = new Date().toISOString();

  poll_title: string = this.poll.poll_title;
  starts_at: string = this.poll.starts_at || this.now;
  ends_at: string = this.poll.ends_at || this.now;
  choices: Choice[] = this.poll.choices;

  get isUpdated(): boolean {
    return (
      !!this.poll_title &&
      this.poll_title !== this.poll.poll_title &&
      this.choices.every(choice => !!choice['choice_text'])
    );
  }

  @Emit('poll-form:submit')
  handleSubmit() {
    return {
      poll_title: this.poll_title,
      choices: this.choices,
      starts_at: this.starts_at,
      ends_at: this.ends_at,
    };
  }
}
</script>

<style lang="scss">
.poll-form__date-time-input {
  display: block;
  font-size: 1rem;
}
</style>
