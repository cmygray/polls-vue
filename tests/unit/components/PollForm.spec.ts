import { createLocalVue, shallowMount } from '@vue/test-utils';
import PollForm from '@/components/PollForm.vue';
import '../setup'
import { poll } from '../__fixtures__';

const localVue = createLocalVue();

describe('PollForm.vue', () => {
  let wrapper: any;

  beforeEach(() => {
    wrapper = shallowMount(PollForm, {
      localVue,
      propsData: {
        poll
      }
    });
  });

  it('renders as expected', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('#handleSubmit', () => {
    const DATA = {
      poll_title: 'POLL TITLE',
      choices: poll.choices.map((choice, index) => ({
        ...choice,
        choice_text: `CHOICE_TEXT ${index + 1}`
      }))
    }
    const EVENT = 'poll-form:submit'

    beforeEach(() => {
      wrapper.setData(DATA)
    })
    it('should emit event with data as expected', () => {
      wrapper.vm.handleSubmit()

      expect(wrapper.emitted()[EVENT]).toEqual(expect.arrayContaining([
        [DATA]
      ]))
    })
  })
});
