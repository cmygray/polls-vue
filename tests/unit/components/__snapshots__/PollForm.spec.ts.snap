// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PollForm.vue renders as expected 1`] = `
<v-form-stub class="poll-form">
  <v-text-field-stub label="Poll Title" value="TEST"></v-text-field-stub>
  <datetime-stub value="2019-06-18T15:28:29.909377" valuezone="Asia/Seoul" inputid="starts_at" inputclass="poll-form__date-time-input" inputstyle="" zone="local" format="yyyy-MM-dd HH:mm" type="datetime" phrases="[object Object]" hourstep="1" minutestep="10" mindatetime="2019-06-18T11:00:00.000Z" weekstart="1"><label for="starts_at">Start at</label></datetime-stub>
  <datetime-stub value="2019-06-21T00:00:00" valuezone="Asia/Seoul" inputid="ends_at" inputclass="poll-form__date-time-input" inputstyle="" zone="local" format="yyyy-MM-dd HH:mm" type="datetime" phrases="[object Object]" hourstep="1" minutestep="10" mindatetime="2019-06-18T11:00:00.000Z" weekstart="1" class="mb-3"><label for="ends_at">End at</label></datetime-stub>
  <v-text-field-stub label="Choice" value="항목을 입력하세요"></v-text-field-stub>
  <v-text-field-stub label="Choice" value="항목을 입력하세요"></v-text-field-stub>
  <v-text-field-stub label="Choice" value="항목을 입력하세요"></v-text-field-stub>
  <v-btn-stub color="primary" activeclass="v-btn--active" tag="button" block="true" type="submit">Save</v-btn-stub>
</v-form-stub>
`;
